fpgaconvnet/models/modules/Module.py

```diff
@@ -8,275 +8,234 @@
+            cls.MODULE_REGISTRY[new_cls.__name__] = new_cls
+        return new_cls
+
+    @classmethod
+    def get_registry(cls):
+        return dict(cls.MODULE_REGISTRY)
+
+    @classmethod
+    def get_all_modules(cls, name: str, backend: str, dimensionality: int):
+
+        # get all the modules in the registry
+        modules = list(cls.MODULE_REGISTRY.values())
+        print(modules)
+
+        # filter all the modules with the given name
+        modules = list(filter(lambda m: m.name == name, modules))
+
+        # filter all the modules with the given backend
+        modules = list(filter(lambda m: m.backend == backend, modules))
+
+        # filter all the modules with the given dimensionality
+        modules = list(filter(lambda m: dimensionality in m.dimensionality, modules))
+
+        return modules
+
+    @classmethod
+    def build(cls, name: str, config: dict, backend: str, dimensionality: int):
+
+        # get all the relevant modules
+        modules = cls.get_all_modules(name, backend, dimensionality)
+
+        # check there is at least 1 module
+        assert len(modules) > 0, f"No modules found for name={name}, \
+                backend={backend}, dimensionality={dimensionality}"
+
+        # check there is only a single module left
+        assert len(modules) == 1, f"Too many modules found for name={name}, \
+                backend={backend}, dimensionality={dimensionality}"
+
+        # get the module class
+        module = modules[0]
+
+        # create a new instance of the module
+        return from_dict(data_class=module, data=config)
+
+    @classmethod
+    def from_config(cls, config: dict):
+        return from_dict(data_class=cls, data=config)
+
+    # @classmethod
+    # @abstractmethod
+    # def generate_random_configuration(cls):
+    #     pass
+
+    # @classmethod
+    # def build_random_inst(cls):
+
+    #     # generate a random configuration
+    #     config = cls.generate_random_configuration()
+
+    #     # create a new instance of the module
+    #     return from_dict(data_class=cls, data=config)
+
+
+@dataclass(kw_only=True)
+class ModuleBase(metaclass=ModuleBaseMeta):
+    name: ClassVar[str]
+    backend: ClassVar[BACKEND]
+    dimensionality: ClassVar[set[DIMENSIONALITY]]
+    register: ClassVar[bool] = False
+    repetitions: int = 1
+
+    @property
+    @abstractmethod
+    def input_ports(self) -> list[Port]:
+        pass
+
+    @property
+    @abstractmethod
+    def output_ports(self) -> list[Port]:
+        pass
+
+    @property
+    @abstractmethod
+    def input_iter_space(self) -> list[list[int]]:
+        pass
+
+    @property
+    @abstractmethod
+    def output_iter_space(self) -> list[list[int]]:
+        pass
+
+    @property
+    @abstractmethod
+    def rate_in(self) -> list[float]:
+        pass
+
+    @property
+    @abstractmethod
+    def rate_out(self) -> list[float]:
+        pass
+
+    @property
+    def ports_in(self) -> int:
```

Comment:
Since this is just the number of ports in and out (below) we might need a different, more indicative, naming convention for that, since the `input_ports()` is really similar and can be really confusing.
I would propose something like `num_input_ports` or `input_ports_size` maybe?